Adopt C++ Concepts in wtf/HashMap

iidmsa opened the pull request:
#### abb93f4101571d5a9188e122277e14de3566782d
<pre>
Adopt C++ Concepts in wtf/HashMap
<a href="https://bugs.webkit.org/show_bug.cgi?id=298008">https://bugs.webkit.org/show_bug.cgi?id=298008</a>
<a href="https://rdar.apple.com/problem/159328083">rdar://problem/159328083</a>

Reviewed by Sam Weinig.

Replace std::enable_if with C++20&apos;s concepts and requires clauses
for template constraints.

* Source/WTF/wtf/CheckedPtr.h:
* Source/WTF/wtf/HashMap.h:
(WTF::M&gt;::find):
(WTF::M&gt;::find const):
(WTF::M&gt;::contains const):
(WTF::M&gt;::inlineGet const):
(WTF::M&gt;::get const):
(WTF::M&gt;::remove):
(WTF::M&gt;::take):

Canonical link: <a href="https://commits.webkit.org/299382@main">https://commits.webkit.org/299382@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03e0e0ec005ce13418f6fca3bdc1b26e4e2f028e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38211 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124702 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70589 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89974 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59526 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; 9 flakes 12 failures; Uploaded test results; 6 flakes 8 failures; Compiled WebKit (cancelled)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121483 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30994 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106279 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70478 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30051 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24390 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68361 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110648 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100433 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24581 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127768 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/117044 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45437 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34279 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98612 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45801 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102500 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98396 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25067 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43824 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21819 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41935 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45307 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50985 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145740 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44770 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37475 "Found 1 new JSC binary failure: testapi, Found 18608 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/StackArgsWithFormals.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48117 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46457 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -198,22 +197,50 @@ class HashMap final {
-    template<typename K = KeyType> typename std::enable_if<IsSmartPtr<K>::value, bool>::type remove(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>*);
-    template<typename K = KeyType> typename std::enable_if<IsSmartPtr<K>::value, MappedTakeType>::type take(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>*);
+    template<typename K = KeyType>
+        requires IsSmartPtrV<K>
```
If we added a `SmartPtr` concept (e.g. `template<typename T> concept SmartPtr = IsSmartPtrV<K>;`) could this be written as:

```c++
    template<SmartPtr K = KeyType>
    iterator find(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>*);
```

iidmsa replied:
Yes. Suggested changes applied.

weinig commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -198,22 +197,50 @@ class HashMap final {
-    template<typename K = KeyType> typename std::enable_if<IsSmartPtr<K>::value && !IsSmartPtr<K>::isNullable, MappedTakeType>::type take(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>&);
+    template<typename K = KeyType>
+        requires (IsSmartPtrV<K> && !IsSmartPtrNullableV<K>)
+    iterator find(std::add_const_t<typename GetPtrHelper<K>::UnderlyingType>&);
```
Same question, but for a NonNullableSmartPtr concept (e.g. `template<typename T> concept NonNullableSmartPtr = SmartPtr<T> && !IsSmartPtrNullableV<T>`).

iidmsa replied:
Yes. Suggested changes applied.

iidmsa committed:
Adopt C++ Concepts in wtf/HashMap
https://bugs.webkit.org/show_bug.cgi?id=298008
rdar://problem/159328083

Reviewed by Sam Weinig.

Replace std::enable_if with C++20's concepts and requires clauses
for template constraints.

* Source/WTF/wtf/CheckedPtr.h:
* Source/WTF/wtf/HashMap.h:
(WTF::M>::find):
(WTF::M>::find const):
(WTF::M>::contains const):
(WTF::M>::inlineGet const):
(WTF::M>::get const):
(WTF::M>::remove):
(WTF::M>::take):

Canonical link: https://commits.webkit.org/299382@main
